### PR TITLE
Stage 4 task completed

### PR DESCRIPTION
**Pull request by MegatronJeremy**
Bonuses were correctly implemented, differences between the server and local game state were eliminated. More GUI additions.

### CODE REVIEW

File: mab/local_game/local_entities/local_entity.py
Comment (irina-scherbitskaya) [inline mab/local_game/local_entities/local_entity.py:1]: why did you rename everything to local?

File: mab/local_game/local_entities/local_entity_enum.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+from enum import StrEnum
+
+
+class LocalEntities(StrEnum):
```
Comment (irina-scherbitskaya) [inline mab/local_game/local_entities/local_entity_enum.py:4]: do you think it is good to store different independent types in one enum?)

File: mab/local_game/local_entities/local_map_features/local_feature.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+from abc import ABC
+
+from local_entities.local_entity import LocalEntity, LocalEntities
+from local_map.local_hex import LocalHex
+
+
+class LocalFeature(LocalEntity, ABC):
```
Comment (irina-scherbitskaya) [inline mab/local_game/local_entities/local_map_features/local_feature.py:7]: why ABC?))

File: mab/local_game/local_entities/local_entity.py
Comment (RicardoSdV) [reply mab/local_game/local_entities/local_entity.py:1]: That is the local version of the game that I'm developing to train the Bot, previously everything had the same name as the other game but that ended up in most of the code disappearing from github for some reason, so I decided It would be better to rename everything to avoid the same problem.

File: mab/local_game/local_entities/local_map_features/local_feature.py (hunk 2 above)
Comment (RicardoSdV) [reply mab/local_game/local_entities/local_map_features/local_feature.py:7]: This is probably a left over from a previous commit, will delete

File: mab/local_game/local_entities/local_map_features/local_landmarks/local_spawn.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+from local_constants import DEFAULT_SPAWN_COLOR
+from local_entities.local_entity_enum import LocalEntities
+from local_entities.local_map_features.local_feature import Feature
+
+
+class LocalSpawn(Feature):
+    def __init__(self, coord: tuple, tank_id: int = -1, color: tuple = DEFAULT_SPAWN_COLOR):
+        self.__belongs_to = tank_id
```
[the remark this answers is not in the document]
Comment (RicardoSdV) [reply mab/local_game/local_entities/local_map_features/local_landmarks/local_spawn.py:8]: OK, will change

File: mab/local_game/local_entities/local_entity.py
Comment (irina-scherbitskaya) [reply mab/local_game/local_entities/local_entity.py:1]: what is the problem with using the main code?

Comment (RicardoSdV) [reply mab/local_game/local_entities/local_entity.py:1]: Well, technically we could, but to fully train the bot I estimate we need between 10 000 to 100 000 games, so any time shaved off the the running of each game makes for either a quicker training or being able to introduce more actions to the bot which would result in a better bot, so being able to run the game locally allows this. Its a complete copy in terms of game logic but it will have no graphics either to save run time on that too.

Comment (RicardoSdV) [reply mab/local_game/local_entities/local_entity.py:1]: Also I will probably just use the parts of the other games' code that will not be changed but rather than introducing a bunch of logic to the changed classes to make the same class run locally and with communication with the server Ill just duplicate those. What do you think of this?

File: main.py [hunk 4]
```diff
@@ -1,5 +1,8 @@
-from gui.display_manager import DisplayManager
+from src.gui.display_manager import DisplayManager
+from mab.train import train
 
 if __name__ == '__main__':
-    display_manager = DisplayManager()
```
Comment (irina-scherbitskaya) [inline main.py:4]: but if I want to run a remote game, do I need to uncomment it?) can you leave the training locally?)

File: src/game.py [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+        self.__next_round: bool = True
+
+        self.__winner: int | None = None
+        self.__winner_index: int | None = None
+        self.__started: bool = False
+        self.__game_is_draw: bool = False
+
+        self.__connection_error: ConnectionError | None = None
+
+        self.__current_turn: list[int] = [-1]
```
Comment (irina-scherbitskaya) [inline src/game.py:34]: better use Dict/List from typing

**Comment by irina-scherbitskaya** [comment]
Now when I start a local game for 3 players, I am catching a bug - "It is not your turn!"(can you check this?).  Are you sure this PR is ready to merge?)

File: src/game.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+
+        if game_state["finished"]:
+            self.__winner = game_state["winner"]
+            if self.__winner is None:
+                self.__game_is_draw = True
+            else:
+                self.__player_wins[self.__winner] += 1
+            self.__print_round_winner()
+
+            if not self.__is_full or game_state["current_round"] == self.__num_rounds:
```
Comment (irina-scherbitskaya) [inline src/game.py:241]: for full game num_rounds = 1, you can check only this

File: main.py (hunk 4 above)
Comment (MegatronJeremy) [reply main.py:4]: Yes, the display manager code only should be run in main, not the training. This was left over from by accident or by a mistake.

File: src/game.py (hunk 5 above)
Comment (MegatronJeremy) [reply src/game.py:34]: What exactly is the rationale behind this?
I looked into it, and by PEP 585: https://peps.python.org/pep-0585/, which is accepted for Python>= 3.9, it says that importing these from typing is deprecated:
slika
Again, if necessary I have no issues with changing these types to the ones from typing, but I was just going by the official Python guidelines.

**Comment by MegatronJeremy** [comment]
This happens when joining a game that is already started, as a non-observer, with the same login parameters. Somebody else was probably running the test game at the same, hence why this happened.  The local game works as it should for what I tested, but I will disallow the previous and generate an error instead.

> Now when I start a local game for 3 players, I am catching a bug - "It is not your turn!"(can you check this?). Are you sure this PR is ready to merge?)

File: src/game.py (hunk 6 above)
Comment (MegatronJeremy) [reply src/game.py:241]: Yes, you are right, thank you.

**Comment by irina-scherbitskaya** [comment]
> This happens when joining a game that is already started, as a non-observer, with the same login parameters. Somebody else was probably running the test game at the same time, hence why this happened. The local game works as it should for what I tested, but I will disallow the previous and generate an error instead.
> 
> > Now when I start a local game for 3 players, I am catching a bug - "It is not your turn!"(can you check this?). Are you sure this PR is ready to merge?)

I give a specific name to the game. I went into the constants and changed the default names for other bots. Each bot made a turn (3 turns passed) and then I get an error.

**Comment by MegatronJeremy** [comment]
> > This happens when joining a game that is already started, as a non-observer, with the same login parameters. Somebody else was probably running the test game at the same time, hence why this happened. The local game works as it should for what I tested, but I will disallow the previous and generate an error instead.
> > > Now when I start a local game for 3 players, I am catching a bug - "It is not your turn!"(can you check this?). Are you sure this PR is ready to merge?)
> 
> I give a specific name for the game. I changed the default names for other bots into the constants . Each bot made a turn (3 turns passed) and then I get an error.

Can you share more of the specifics? I really haven't been able to reproduce this error.

**Comment by irina-scherbitskaya** [comment]
if I noticed correctly, you have a constant color vehicles for spawns during a full game (not related to players), but maybe it should be

**Comment by MegatronJeremy** [comment]
> > > > This happens when joining a game that is already started, as a non-observer, with the same login parameters. Somebody else was probably running the test game at the same time, hence why this happened. The local game works as it should for what I tested, but I will disallow the previous and generate an error instead.
> > > > > Now when I start a local game for 3 players, I am catching a bug - "It is not your turn!"(can you check this?). Are you sure this PR is ready to merge?)
> > > 
> > > 
> > > I give a specific name for the game. I changed the default names for other bots into the constants . Each bot made a turn (3 turns passed) and then I get an error.
> > 
> > 
> > Can you share more of the specifics? I really haven't been able to reproduce this error.
> 
> So, I understand the problem. You are extracting a buffer_size response once. But in fact, it may not be completely extracted at one time and less data will get. Here you can read separately - the first 4 bytes as a response code, the next 4 bytes as the length of the message. Then read the message in a loop with some chunks by sock.recv (read each time 4096 bytes, but this value can be less - 1024, 2048.) until size < msg size.

Thank you for noticing this! Yes, it was a problem, I tested every packet received manually as well, and not everything was received at a few points over multiple tests. This somehow wasn't an issue for any of us (my guess is that most of the data we used was just received correctly, or that this just happened too sporadically), but I guess it depends on connectivity and a lot of different factors. Now it should always be correct regardless of the underlying hardware, I basically made the adjustments as you said.

**Comment by wanjoh** [comment]
> if I noticed correctly, you have a constant color vehicles for spawns during a full game (not related to players), but maybe it should be

Sorry but I don't think I understand what exactly do you mean? Both spawn color and tank color are the same as player's color and in case of a full game, there are 6 rounds where spawn points move to different place.

**Comment by irina-scherbitskaya** [comment]
> > if I noticed correctly, you have a constant color vehicles for spawns during a full game (not related to players), but maybe it should be
> 
> Sorry but I don't think I understand what exactly do you mean? Both spawn color and tank color are the same as player's color and in case of a full game, there are 6 rounds where spawn points move to different place.

Maybe I was looking at revisions where it didn't work yet.
